### PR TITLE
[Feature] Support Denoising Unet for DDPM-based models

### PR DESCRIPTION
**Pull request by LeoXing1996**

### CODE REVIEW

File: mmgen/models/ddpm/denoising.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+from .modules import EmbedSequential
+
+
+@MODULES.register_module()
+class DenoisingUnet(nn.Module):
+    """Denoising Unet. This network receives a diffused image ``x_t`` and
+    current timestep ``t``, and returns a ``output_dict`` corresponding to the
+    passed ``output_cfg``.
+
+    ``output_cfg`` define the number of channels and the meaning of the output.
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:21]: define ---> defines

File: mmgen/models/ddpm/denoising.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+    ``output_cfg`` support ``mean`` and ``var`` for keys, and
+    denotes how the the network output mean and variance required for denoising
+    process.
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:24]: ``output_cfg`` mainly contains keys of ``mean`` and ``var``, denoting how the network outputs mean and variance required for the denoising process.

File: mmgen/models/ddpm/denoising.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    For ``mean``:
+    1. ``dict(mean='EPS')``: Model would predict noise added in the
+        diffusion process. And the ``output_dict`` would contains a key named
+        ``eps_t_pred``.
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:28]: contains --> contain (Please also check other places)

File: mmgen/models/ddpm/denoising.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+    2. ``dict(mean='START_X')``: Model would direct the mean of the original
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:29]: What's the meaning of `direct`?

File: mmgen/models/ddpm/denoising.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        image `x_0`. And the ``output_dict`` would contains a key named
+        ``x_0_pred``.
+    3. ``dict(mean='X_TM1_PRED')``: Model would predict the mean of diffused
+        image at `t-1` timestep. And the ``output_dict`` would contains a key
+        named ``x_tm1_pred``.
+
+    For ``var``:
+    1. ``dict(var='FIXED_SMALL')`` or ``dict(var='FIXED_LARGE')``: Variance in
+        the denoising process is regarded as a fixed value. The output of
+        network would only have three channels.
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:39]: Why do you emphasize the three channels? What's the general case of the output channels?

File: mmgen/models/ddpm/denoising.py [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+    2. ``dict(var='LEARNED')``: Model would predict `log_variance` in the
+        denoising process. And the ``output_dict`` would contains a key named
+        ``log_var``.
+    3. ``dict(var='LEARNED_RANGE')``: Model would predict an interpolation
+        factor and the `log_variance` would be calculated as
+        `factor * upper_bound + (1-factor) * lower_bound`. And the
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:45]: No `And`, here.

File: mmgen/models/ddpm/denoising.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        ``output_dict`` would contains a key named ``factor``.
+
+    If ``FIXED`` not in ``var``, the number of output channels would be
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:48]: ``FIXED`` not in ``var``  --->  ``var`` is not ``'FIXED_SMALL'``

File: mmgen/models/ddpm/denoising.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+    the double of input channels. Otherwise, the number of output channels
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:49]: the double of input channels, where the first half part contains predicted mean values and the other part is the predicted variance values.

File: mmgen/models/ddpm/denoising.py [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+    equals to the input channels.
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:50]: ... the input channels, only containing the predicted mean values.

File: mmgen/models/ddpm/denoising.py [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+        elif isinstance(channels_cfg_, list):
+            self.channel_factor_list = channels_cfg_
+        else:
+            raise ValueError('Only support list or dict for `channel_cfg`, '
+                             f'receive {type(channels_cfg_)}')
+
+        embedding_channels = base_channels * 4 \
+            if embedding_channels == -1 else embedding_channels
+        self.time_embedding = build_module(
+            dict(type='TimeEmbedding'), {
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:194]: If you perform a hard-code design, you may directly define an instance of the `TimeEmbedding` instead of using `build_module`.

File: mmgen/models/ddpm/denoising.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+        self.resblock_cfg.setdefault('dropout', dropout)
+        self.resblock_cfg.setdefault('norm_cfg', norm_cfg)
+        self.resblock_cfg.setdefault('act_cfg', act_cfg)
+        self.resblock_cfg.setdefault('embedding_channels', embedding_channels)
+        self.resblock_cfg.setdefault('use_scale_shift_norm',
+                                     use_scale_shift_norm)
+        self.resblock_cfg.setdefault('shortcut_kernel_size',
+                                     shortcut_kernel_size)
+
+        # factor to apply attention
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:212]: What's the meaning of `factor to apply attention`?

File: mmgen/models/ddpm/denoising.py [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+        self.attention_cfg = deepcopy(attention_cfg)
+        self.attention_cfg.setdefault('num_heads', num_heads)
+        self.attention_cfg.setdefault('norm_cfg', norm_cfg)
+
+        self.downsample_cfg = deepcopy(downsample_cfg)
+        self.downsample_cfg.setdefault('with_conv', downsample_conv)
+        self.upsample_cfg = deepcopy(upsample_cfg)
+        self.upsample_cfg.setdefault('with_conv', upsample_conv)
+
+        # init the factor
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:223]: Do you mean `channel scale factor`? We should give clear hints or comments.

File: mmgen/models/ddpm/denoising.py [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+        scale = 1
+        self.in_blocks = nn.ModuleList([
+            EmbedSequential(
+                nn.Conv2d(in_channels, base_channels, 3, 1, padding=1))
+        ])
+        self.in_channels_list = [base_channels]
+        for level, factor in enumerate(self.channel_factor_list):
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:230]: Is this `for-loop` for the encoder part in U-Net? If it is, leave a comment: `#  construct the encoder part`.

File: mmgen/models/ddpm/denoising.py [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+
+            if level != len(self.channel_factor_list) - 1:
+                self.in_blocks.append(
+                    EmbedSequential(
+                        build_module(self.downsample_cfg,
+                                     {'in_channels': in_channels_})))
+                self.in_channels_list.append(in_channels_)
+                scale *= 2
+
+        self.mid_blocks = EmbedSequential(
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:260]: Offer a comment here to clarify what this part is for.

File: mmgen/models/ddpm/denoising.py [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+            build_module(self.resblock_cfg, {'in_channels': in_channels_}),
+            build_module(self.attention_cfg, {'in_channels': in_channels_}),
+            build_module(self.resblock_cfg, {'in_channels': in_channels_}),
+        )
+
+        in_channels_list = deepcopy(self.in_channels_list)
```
Comment (nbei) [inline mmgen/models/ddpm/denoising.py:266]: Leave a comment here to clarify what you will do in the next part.

File: mmgen/models/ddpm/denoising.py (hunk 5 above)
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:39]: Three channels for  RGB image.

File: mmgen/models/ddpm/denoising.py [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+            generator. The other layers contains channels based on this number.
+            Defaults to ``128``.
+        resblocks_per_downsample (int, optional): Number of ResBlock used
+            between two downsample operations. And number of ResBlock between
+            upsample operations would be the same value to keep symmetry.
+            Defaults to 3.
+        num_timesteps (int, optional): The total timestep of the denoising
+            process and the diffusion process. Defaults to ``1000``.
+        rescale_timesteps (bool, optional): Whether rescale the input timestep
+            in range of [0, 1000].  Defaults to ``True``.
```
[the remark this answers is not in the document]
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:66]: Yes, this range is fixed in [improve-DDPM](https://github.com/openai/improved-diffusion/blob/e94489283bb876ac1477d5dd7709bbbd2d9902ce/improved_diffusion/gaussian_diffusion.py#L353).

File: mmgen/models/ddpm/denoising.py [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+        dropout (float, optional): The probability of dropout operation of
+            each ResBlock. Pass ``0`` to do not use dropout. Defaults as 0.
```
[the remark this answers is not in the document]
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:68]: In both our and official implementation, dropout is used in each ResBlock.

File: mmgen/models/ddpm/denoising.py [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+                image_size) == 2, 'The length of `image_size` should be 2.'
+            assert image_size[0] == image_size[
+                1], 'Width and height of the image should be same.'
+            image_size = image_size[0]
+        self.image_size = image_size
+
+        channels_cfg_ = deepcopy(self._default_channels_cfg)
+        if channels_cfg is not None:
+            channels_cfg_ = channels_cfg_.update(channels_cfg)
+        if isinstance(channels_cfg_, dict):
```
[the remark this answers is not in the document]
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:179]: In line 176, we initialize `channels_cfg_` with `self._default_channels_cfg`, which is an instance of dict.

File: mmgen/models/ddpm/denoising.py (hunk 13 above)
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:235]: Comment has added.

File: mmgen/models/ddpm/denoising.py (hunk 14 above)
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:266]: Comment has been added.

File: mmgen/models/ddpm/denoising.py (hunk 15 above)
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:273]: Comment has been added.

File: mmgen/models/ddpm/denoising.py (hunk 3 above)
Comment (LeoXing1996) [reply mmgen/models/ddpm/denoising.py:28]: Revised.

**Review by nbei — APPROVED** [review]
